presto-hive/src/main/java/io/prestosql/plugin/hive/RegisterPartitionProcedure.java

```diff
@@ -0,0 +1,173 @@
+    private final HdfsEnvironment hdfsEnvironment;
+    private final HiveMetastoreClosure metastore;
+
+    @Inject
+    public RegisterPartitionProcedure(Supplier<TransactionalMetadata> hiveMetadataFactory, HiveMetastore metastore, HdfsEnvironment hdfsEnvironment)
+    {
+        this.hiveMetadataFactory = requireNonNull(hiveMetadataFactory, "hiveMetadataFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.metastore = new HiveMetastoreClosure(requireNonNull(metastore, "metastore is null"));
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "register_partition",
+                ImmutableList.of(
+                        new Procedure.Argument("schema_name", VARCHAR),
+                        new Procedure.Argument("table_name", VARCHAR),
+                        new Procedure.Argument("location", VARCHAR),
+                        new Procedure.Argument("partition_columns", new ArrayType(VARCHAR)),
+                        new Procedure.Argument("partition_values", new ArrayType(VARCHAR))),
+                REGISTER_PARTITION.bindTo(this));
+    }
+
+    public void registerPartition(ConnectorSession session, String schemaName, String tableName, String location, List<Object> partitionColumnNames, List<Object> partitionValues)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            doRegisterPartition(session, schemaName, tableName, location, partitionColumnNames, partitionValues);
+        }
+    }
+
+    private void doRegisterPartition(ConnectorSession session, String schemaName, String tableName, String location, List<Object> partitionColumnNames, List<Object> partitionValues)
+    {
+        HiveIdentity identity = new HiveIdentity(session);
+        HdfsContext hdfsContext = new HdfsContext(session, schemaName, tableName);
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+
+        Path partitionLocation = new Path(location);
+
+        Table table = metastore.getTable(identity, schemaName, tableName)
+                .orElseThrow(() -> new TableNotFoundException(schemaTableName));
+
+        if (table.getTableType().equals(TableType.VIRTUAL_VIEW.name())) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Table is a view: " + schemaTableName);
+        }
+
+        if (table.getPartitionColumns().isEmpty()) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Table is not partitioned: " + schemaTableName);
+        }
+
+        List<String> actualPartitionColumnNames = table.getPartitionColumns().stream()
+                .map(Column::getName)
+                .collect(toImmutableList());
+
+        if (!Objects.equals(partitionColumnNames, actualPartitionColumnNames)) {
```

Comment:
I think the comparison should be case insensitive.

Reply:
I'll make a follow-up PR changing this in every procedure (create empty partition as well)